### PR TITLE
wallet: return change from SelectionResult

### PR DESCRIPTION
**Pull request by S3RK**
Benefits:
1. more accurate waste calculation for knapsack. Waste calculation is now consistent with tx building code. Before we always assumed change for knapsack even when the solution is changeless4. 
2. simpler tx building code. Only create change output when it's needed
3. makes it easier to correctly account for fees for CPFP inputs (should be done in a follow up) 

In the first three commits we fix the code to accurately track selection target in `SelectionResult::m_target`
Then we introduce new variable `min_change` that represents the minimum viable change amount
Then we introduce `SelectionResult::GetChange()` which incapsulates dropping change for fee logic and uses correct values of `SelectionResult::m_target`
Then we use `SelectionResult::GetChange()` in both tx building and waste calculation code

This PR is a refactoring and shouldn't change the behaviour.
There is only one known small change (arguably a bug fix). Before we dropped change output if it's smaller than `cost_of_change` after paying change fees. This is incorrect as `cost_of_change` already includes `change_fee`.

### CODE REVIEW

File: src/wallet/coinselection.cpp [hunk 1]
```diff
@@ -156,7 +156,7 @@ std::optional<SelectionResult> SelectCoinsBnB(std::vector<OutputGroup>& utxo_poo
     for (const size_t& i : best_selection) {
         result.AddInput(utxo_pool.at(i));
     }
-    result.ComputeAndSetWaste(CAmount{0});
+    result.ComputeAndSetWaste(cost_of_change, cost_of_change, CAmount{0});
```
Comment (Xekyo) [inline src/wallet/coinselection.cpp:159]: In "wallet: use GetChange() when computing waste" (45ee9d3a737cc23babb9200c4d75d7dae4dc6700):

Shouldn't the first value here be the maximum of `cost_of_change`  and  `dust_threshold`? I think that the former can be lower than the latter for feerates below discard_feerate.

File: src/wallet/coinselection.cpp [hunk 2]
```diff
@@ -406,9 +406,15 @@ CAmount GenerateChangeTarget(CAmount payment_value, FastRandomContext& rng)
     }
 }
 
-void SelectionResult::ComputeAndSetWaste(CAmount change_cost)
+void SelectionResult::ComputeAndSetWaste(const CAmount min_change, const CAmount change_cost, const CAmount change_fee)
```
Comment (Xekyo) [inline src/wallet/coinselection.cpp:409]: In "wallet: use GetChange() when computing waste" (45ee9d3a737cc23babb9200c4d75d7dae4dc6700):

As suggested before, I'd prefer `min_viable_change` for this value here and below.

File: src/wallet/spend.cpp [hunk 3]
```diff
@@ -745,6 +745,13 @@ static BResult<CreatedTransactionResult> CreateTransactionInternal(
     coin_selection_params.m_change_fee = coin_selection_params.m_effective_feerate.GetFee(coin_selection_params.change_output_size);
     coin_selection_params.m_cost_of_change = coin_selection_params.m_discard_feerate.GetFee(coin_selection_params.change_spend_size) + coin_selection_params.m_change_fee;
 
+    // The smallest change amount should be:
+    // 1. at least equal to dust threshold
+    // 2. at least 1 sat greater than fees to spend it at m_discard_feerate
+    const auto dust = GetDustThreshold(change_prototype_txout, coin_selection_params.m_discard_feerate);
+    const auto change_spend_fee = coin_selection_params.m_discard_feerate.GetFee(coin_selection_params.change_spend_size);
+    coin_selection_params.min_change = std::max(change_spend_fee + 1, dust);
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/spend.cpp:753]: I'm not sure I understand the question about the remainder. I've tried to keep the existing behaviour unchanged.

We can totally add +1 to the dust threshold but that would be a change in behaviour. Are we sure we want to do it?

File: src/wallet/spend.cpp [hunk 4]
```diff
@@ -792,23 +792,23 @@ static BResult<CreatedTransactionResult> CreateTransactionInternal(
...
+            nChangePosInOut = rng_fast.randrange(txNew.vout.size() + 1);
+        }
+        else if ((unsigned int)nChangePosInOut > txNew.vout.size())
+        {
+            return _("Transaction change output index out of range");
+        }
+        txNew.vout.insert(txNew.vout.begin() + nChangePosInOut, newTxOut);
+    } else {
+        nChangePosInOut = -1;
     }
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/spend.cpp:810]: I think I fixed it

File: src/wallet/coinselection.h [hunk 5]
```diff
@@ -123,6 +123,10 @@ struct CoinSelectionParams {
     /** Mininmum change to target in Knapsack solver: select coins to cover the payment and
      * at least this value of change. */
     CAmount m_min_change_target{0};
+    /** Minimum amount for creating a change output.
+     * If change budget is smaller than min_change then we forgo creation of change output.
+     */
+    CAmount min_change{0};
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/coinselection.h:129]: done

File: src/wallet/coinselection.cpp [hunk 6]
```diff
@@ -480,4 +485,24 @@ std::string GetAlgorithmName(const SelectionAlgorithm algo)
...
+    const CAmount change_budget = m_use_effective
+                                  ? GetSelectedEffectiveValue() - m_target - change_fee
+                                  : GetSelectedValue() - m_target;
+
+    if (change_budget < min_change) {
+        return 0;
+    }
+
+    return change_budget;
+}
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/coinselection.cpp:512]: done

File: src/wallet/coinselection.cpp (hunk 2 above)
Comment (S3RK) [reply src/wallet/coinselection.cpp:409]: done

File: src/wallet/coinselection.cpp (hunk 1 above)
Comment (S3RK) [reply src/wallet/coinselection.cpp:159]: Technically yes, but it doesn't matter in this case because BnB would never have change more than `cost_of_change`.

I'd say a better way to fix it is to use `min_viable_change` for the upper bound of BnB. But I don't want to do expand the scope of this PR. Can we leave it for the follow up?

File: src/wallet/spend.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/spend.cpp]: The changeless window is based on `cost_of_change`, not just `change_spend_fee`. So the minimum viable change should be `cost_of_change + 1`.

I agree that the `+ 1` should not be applied to the dust threshold. The existing behavior throws away change when it is `< dust` or `<= cost of change`. So that's equivalent to `< dust` or `< cost of change + 1`.

[the remark this answers is not in the document]
Comment (Xekyo) [reply src/wallet/spend.cpp]: Err, yes. You're right, the `+1` should be on `coin_selection_params.m_cost_of_change`

[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/spend.cpp]: No. `cost_of_change` = `change_fee` + `change_spend_fee`. When we compare `change` To `min_viable_change` the `change_fee` is already deducted.

Think about it from this perspective, you don't want to create a change output that is less than what it takes to spend it.

[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/spend.cpp]: You're right thought. That that right now we drop change smaller than `cost_of_change` after deducting the `change_fee`. So the `change_fee` is deducted twice. Which I think is a bug. But I need to disclose this in the description.

[the remark this answers is not in the document]
Comment (achow101) [reply src/wallet/spend.cpp]: Ah ok. That makes sense.

**Comment by achow101** [comment]
ACK e763496caab2f9b9f9ba038f7a020caa56997c1b

File: src/wallet/spend.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (Xekyo) [reply src/wallet/spend.cpp]: Okay, convinced that all sounds right. Thanks for explaining

File: src/wallet/coinselection.cpp (hunk 1 above)
Comment (Xekyo) [reply src/wallet/coinselection.cpp:159]: Ah right, and since this call originates from `SelectCoinsBnB(…)` that's sufficient of course.

File: src/wallet/spend.cpp [hunk 7]
```diff
@@ -489,31 +489,19 @@ std::optional<SelectionResult> ChooseSelectionResult(const CWallet& wallet, cons
...
-    std::vector<OutputGroup> all_groups = GroupOutputs(wallet, available_coins, coin_selection_params, eligibility_filter, false /* positive_only */);
-    CAmount target_with_change = nTargetValue;
-    // While nTargetValue includes the transaction fees for non-input things, it does not include the fee for creating a change output.
-    // So we need to include that for KnapsackSolver and SRD as well, as we are expecting to create a change output.
-    if (!coin_selection_params.m_subtract_fee_outputs) {
-        target_with_change += coin_selection_params.m_change_fee;
-    }
-    if (auto knapsack_result{KnapsackSolver(all_groups, target_with_change, coin_selection_params.m_min_change_target, coin_selection_params.rng_fast)}) {
+    std::vector<OutputGroup> all_groups = GroupOutputs(wallet, available_coins, coin_selection_params, eligibility_filter, /*positive_only=*/false);
+    if (auto knapsack_result{KnapsackSolver(all_groups, nTargetValue, coin_selection_params.m_min_change_target, coin_selection_params.rng_fast)}) {
```
Comment (Xekyo) [inline src/wallet/spend.cpp:499]: Just a headscratcher: 
In the commit message of “wallet: accurate SelectionResult::m_target” (98d5dd26aec30ec63b555afc844281a4a04bd0fb) you write:

> SelectionResult::m_target should be equal to actual selection target.
> Selection target is the sum of all recipient amounts plus non input fees.
> So we need to remove change_fee from the m_target. It's safe because change
> target is always greater than the change fee, so we can always cover fees
> if change output is created.

It's not clear to me why the change target is always greater than change fee. Are you just alluding to the minimum change being more than the change fee even beyond 1,000 s/vB, or does the change fee actually get added anywhere that I'm missing?

File: src/wallet/coinselection.h [hunk 8]
```diff
@@ -123,6 +123,10 @@ struct CoinSelectionParams {
     /** Mininmum change to target in Knapsack solver: select coins to cover the payment and
      * at least this value of change. */
     CAmount m_min_change_target{0};
+    /** Minimum amount for creating a change output.
+     * If change budget is smaller than min_change then we forgo creation of change output.
+     */
+    CAmount min_viable_change{0};
```
Comment (Xekyo) [inline src/wallet/coinselection.h:129]: Nit: 
In wallet: calculate and store min_change (84f871db7bc2e74c5b9ebdd4690a9eb5c79ef334) the change introduces now `min_viable_change` but the commit message still says `min_change`.

**Review by murchandamus — COMMENTED** [review]
ACK e763496caab2f9b9f9ba038f7a020caa56997c1b

File: src/wallet/spend.cpp (hunk 7 above)
Comment (S3RK) [reply src/wallet/spend.cpp:499]: Good catch. Indeed what I meant is that `m_change_target` is at least `CHANGE_LOWER` which in most cases should be bigger than `change_fee`, but at high enough fee rate it doesn't hold.

Maybe we can ensure that `m_change_target` is at least `change_fee` by making `CHANGE_LOWER` dynamic and equal to max(50000, change_fee)`. @Xekyo @achow101 wdyt?

File: src/wallet/coinselection.h (hunk 8 above)
Comment (S3RK) [reply src/wallet/coinselection.h:129]: ack. I'll fix it on the next push

File: src/wallet/spend.cpp (hunk 7 above)
Comment (S3RK) [reply src/wallet/spend.cpp:499]: @furszy #25825

File: src/wallet/coinselection.h (hunk 8 above)
Comment (S3RK) [reply src/wallet/coinselection.h:129]: done

File: src/wallet/spend.cpp (hunk 7 above)
Comment (S3RK) [reply src/wallet/spend.cpp]: I've added c8cf08ea743e430c2bf3fe46439594257b0937e5 to address this. Please review

**Comment by S3RK** [comment]
Added one commit c8cf08ea743e430c2bf3fe46439594257b0937e5 and rebased on master again

File: src/wallet/spend.cpp [hunk 9]
```diff
@@ -855,6 +855,13 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
 
     coin_selection_params.m_min_change_target = GenerateChangeTarget(std::floor(recipients_sum / vecSend.size()), coin_selection_params.m_change_fee, rng_fast);
 
+    // The smallest change amount should be:
+    // 1. at least equal to dust threshold
+    // 2. at least 1 sat greater than fees to spend it at m_discard_feerate
+    const auto dust = GetDustThreshold(change_prototype_txout, coin_selection_params.m_discard_feerate);
+    const auto change_spend_fee = coin_selection_params.m_discard_feerate.GetFee(coin_selection_params.change_spend_size);
+    coin_selection_params.min_viable_change = std::max(change_spend_fee + 1, dust);
```
Comment (furszy) [inline src/wallet/spend.cpp:863]: hmm, this doesn't seems to be right:

1) The dust amount is taken from --> `GetDustThreshold` which receives the change output, serializes it and uses magic numbers to calculate the input vsize (it's a general minimum input vsize calculation for an output), then calls `discard_fee.Get(vsize)` with it to get the dust amount.

2. The `change_spend_fee` is taken from --> `discard_fee.Get(change_spend_size)` , where `change_spend_size` is the result of `CalculateMaximumSignedInputSize(output)` which is the real vsize calculation of a signed input that spends the change output (we don't hardcode numbers here, we create the input and sign it).


In other words:

The only difference that I'm seeing between the `dust` and `change_spend_fee` variables is the vsize calculation (both are doing `m_discard_fee.GetFee(vsize)`), and:

- for `dust`, vsize is calculated from hardcoded numbers to get the minimum theoretical possible size.
- for `change_spend_fee`, vsize is calculated creating an input and dummy signing it.

So, `dust` will always be smaller than `change_spend_fee` right?

Comment (furszy) [reply src/wallet/spend.cpp:863]: Extra note, discussed this topic with @theStack for a while. He mentioned that you might wanted to use `wallet.chain().relayDustFee()` instead of `coin_selection_params.m_discard_feerate` in the dust amount calculation?

Comment (achow101) [reply src/wallet/spend.cpp:863]: What's implemented here matches the behavior in master. For this PR, I think this is correct.

However we should probably use `m_long_term_feerate` for calculating the change spend fee rather than the discard feerate. This would make it match how we calculate "future spends" for all of our inputs.

> you might wanted to use `wallet.chain().relayDustFee()` instead of `coin_selection_params.m_discard_feerate` in the dust amount calculation?

I think discard feerate is correct here. It is the maximum of the longest term fee estimate and the dust relay fee. As the name suggests, it is the feerate that we are willing to discard at, so I think it makes sense to keep it as is.

Comment (theStack) [reply src/wallet/spend.cpp:863]: > > you might wanted to use `wallet.chain().relayDustFee()` instead of `coin_selection_params.m_discard_feerate` in the dust amount calculation?
> 
> I think discard feerate is correct here. It is the maximum of the longest term fee estimate and the dust relay fee. As the name suggests, it is the feerate that we are willing to discard at, so I think it makes sense to keep it as is.

Isn't the point of setting the smallest change amount to "at least equal to dust threshold"  to ensure that the created tx can enter our own mempool now, rather than somewhen in the future (considering that we run a node with a possibly customized `-dustrelayfee=...` option)? As an example, if one would set `-dustrelayfee=...` to a high value and `discardfee=0`, a tx with a change output could be created that is rejected from our mempool (due to being `IsDust` and therefore `!IsStandardTx()`) and thus can't be sent. That's why my assumption was that we want to call `GetDustThreshold` with the dustRelayFee instead:

```diff
-    const auto dust = GetDustThreshold(change_prototype_txout, coin_selection_params.m_discard_feerate);
+    const auto dust = GetDustThreshold(change_prototype_txout, wallet.chain.relayDustFee());
```
(Am I missing something?)

Comment (achow101) [reply src/wallet/spend.cpp:863]: > As an example, if one would set `-dustrelayfee=...` to a high value and `discardfee=0`, a tx with a change output could be created that is rejected from our mempool

Then the dustrelayfee is what `m_discard_fee` is set to. See `GetDiscardRate` in src/wallet/fees.cpp.

**Review by furszy — APPROVED** [review]
Code review ACK 4fef5344

Great step forward, this area needs so much more love.

**Comment by achow101** [comment]
ACK 4fef5344288e454460b80db0316294e1ec1ad8ad

File: src/wallet/spend.cpp (hunk 9 above)
Comment (theStack) [reply src/wallet/spend.cpp:863]: > > As an example, if one would set `-dustrelayfee=...` to a high value and `discardfee=0`, a tx with a change output could be created that is rejected from our mempool
> 
> Then the dustrelayfee is what `m_discard_fee` is set to. See `GetDiscardRate` in src/wallet/fees.cpp.

Oh right, I completely overlooked this following line of code, d'oh! 🤦‍♂️ 

https://github.com/bitcoin/bitcoin/blob/a8f69541ad53a76d5f69044ba829069d225a4af1/src/wallet/fees.cpp#L91

**Comment by theStack** [comment]
Concept ACK

**Review by w0xlt — COMMENTED** [review]
Approach ACK.

**Review by w0xlt — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/25647/commits/4fef5344288e454460b80db0316294e1ec1ad8ad

File: src/wallet/spend.cpp (hunk 9 above)
Comment (S3RK) [reply src/wallet/spend.cpp:863]: I believe we should get rid of `discard fee rate` and replace it with `dust relay fee rate` and `long term fee rate`. For this PR I decide to stick to the current implementation to limit the scope. I'll open an issue to deprecate `discard fee rate` if there is none yet.

File: src/wallet/spend.cpp [hunk 10]
```diff
@@ -607,12 +607,15 @@ std::optional<SelectionResult> SelectCoins(const CWallet& wallet, CoinsResult& a
...
+    SelectionResult preselected(preset_inputs.GetSelectionAmount(), SelectionAlgorithm::MANUAL);
+    preselected.AddInput(preset_inputs);
+
     // Coin Selection attempts to select inputs from a pool of eligible UTXOs to fund the
     // transaction at a target feerate. If an attempt fails, more attempts may be made using a more
     // permissive CoinEligibilityFilter.
     std::optional<SelectionResult> res = [&] {
         // Pre-selected inputs already cover the target amount.
-        if (value_to_select <= 0) return std::make_optional(SelectionResult(nTargetValue, SelectionAlgorithm::MANUAL));
+        if (value_to_select <= 0) return std::make_optional(SelectionResult(value_to_select, SelectionAlgorithm::MANUAL));
```
Comment (Xekyo) [inline src/wallet/spend.cpp:618]: In “wallet: account for preselected inputs in target” (e3210a722542a9cb5f7e4be72470dbe488c281fd):

Sorry, if I'm being dense, but how was “deducting the effective values of the preseleted inputs from `m_target`” leading to a different result than “creating a separate SelectionResult with the target set to the effective values of the pre-selected inputs and then starting the selection targeting the delta between that and the search target”?

It sounds to me that in both cases that would be `m_target - effective_value_of_preselection` under the hood? I assume there is a subtle difference here, but it's not obvious to me from the description. Is the difference that once we use `m_target` and “search target” is now composed differently? Please feel free to ignore if this has been covered elsewhere, and I just missed it.

**Review by murchandamus — COMMENTED** [review]
Sorry for the slow re-review. I don't think these two are blockers if they're issues at all, but I figured it might be good to bring them up.

File: src/wallet/coinselection.cpp [hunk 11]
```diff
@@ -480,4 +485,24 @@ std::string GetAlgorithmName(const SelectionAlgorithm algo)
...
+{
+    // change = SUM(inputs) - SUM(outputs) - fees
+    // 1) With SFFO we don't pay any fees
+    // 2) Otherwise we pay all the fees:
+    //  - input fees are covered by GetSelectedEffectiveValue()
+    //  - non_input_fee is included in m_target
+    //  - change_fee
+    const CAmount change = m_use_effective
+                           ? GetSelectedEffectiveValue() - m_target - change_fee
+                           : GetSelectedValue() - m_target;
```
[the remark this answers is not in the document]
Comment (S3RK) [reply src/wallet/coinselection.cpp:505]: With SFFO `non_input_fee` is always set to 0. We can probably make this simpler and cleaner, let's discuss it on the wallet meeting or in some other place.

File: src/wallet/spend.cpp (hunk 10 above)
Comment (achow101) [reply src/wallet/spend.cpp:618]: The newly added `Merge` function will add the `m_target`s of the two `SelectionResults`. In the event that `value_to_select <= 0`, using `nTargetValue` in this `SelectionResult will add `nTargetValue` to the `m_target` stored in the preset inputs `SelectionResult`. This is incorrect because the target was already met and we're actually just adding it twice.

Also note how all of the other `SelectionResult`s returned by this function all use `value_to_select`.

File: src/wallet/coinselection.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (Xekyo) [reply src/wallet/coinselection.cpp:505]: Ah okay, that explains it. Thanks

File: src/wallet/spend.cpp (hunk 10 above)
Comment (Xekyo) [reply src/wallet/spend.cpp:618]: Answering my own question after talking to S3RK out of band:

The correct `m_target` is needed to calculate the change in advance. Previously, the preselection approach was reducing `m_target` to reflect that some funds had already been picked, and now storing them in a SelectionResult and combining the two before calculating the change will be operating on basis of the correct `m_target` instead.

**Comment by Xekyo** [comment]
crACK 4fef5344288e454460b80db0316294e1ec1ad8ad

File: src/wallet/spend.cpp (hunk 10 above)
Comment (Xekyo) [reply src/wallet/spend.cpp:618]: Ah thanks, @achow101, the case with the negative target makes it even clearer how this is better.

Comment (furszy) [reply src/wallet/spend.cpp:618]: Small extra add to the topic in case you haven't seen it; the pre-set inputs workflow (this stuff) is being decoupled from the coin selection process in #25685. Which should make the flow clearer and less error-prone.
